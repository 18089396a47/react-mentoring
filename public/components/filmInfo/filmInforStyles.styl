.film-info
  display: flex;
  padding: 0 40px 15px;

  &__poster
    width: 220px;
    margin-right: 60px;

  &__info-wrapper
    display: flex;
    flex-direction: column;
    padding: 25px 0;
    color: #d0d0d0;

  &__title-wrapper
    display: flex;
    align-items: center;
    margin-bottom: 5px;

  &__title
    font-weight: bold;
    font-size: 26px;
    color: #f45162;

  &__rating
    font-size: 18px;
    color: #a9d67c;
    border: 1px solid #a9d67c;
    border-radius: 50%;
    width: 40px;
    height: 40px;
    line-height: 40px;
    text-align: center;
    margin-left: 20px;

  &__year-wrapper
    display: flex;
    color: #fff;
    font-weight: bold;
    margin-top: 15px;

  &__year
    margin-right: 30px;

  &__summary
    margin-top: 35px;
    width: 460px;

  &__director
    margin-top: 30px;
    font-size: 12px;

  &__cast
    margin-top: 15px;
    font-size: 12px;
    width: 460px;
