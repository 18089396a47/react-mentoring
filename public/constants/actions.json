{
  "searchStart": "SEARCH_START",
  "searchEnd": "SEARCH_END",
  "changeSearchInput": "CHANGE_SEARCH_INPUT",
  "updateSearchResults": "UPDATE_SEARCH_RESULTS"
}
