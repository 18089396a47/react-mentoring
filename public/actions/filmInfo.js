import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const searchFilmEnd = () => ({
    type: types.searchFilmEnd
});

export const searchFilmStart = (id, searchType) => ({
    type: types.searchFilmStart,
    id,
    searchType
});

function* fetchFilm({ id, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.movies.getById : theMovieDb.tv.getById;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            id
        }));

        yield put(searchFilmEnd());
        yield put(updateFilmInfo(JSON.parse(response)))
    } catch (e) {
        yield put(searchFilmEnd());
    }
}

function* watchFetchFilm() {
    yield takeLatest(types.searchFilmStart, fetchFilm);
}

export function* filmInfoSaga() {
    yield all([
        watchFetchFilm()
    ]);
}

export const updateFilmInfo = (data) => ({
    type: types.updateFilmInfo,
    data
});
