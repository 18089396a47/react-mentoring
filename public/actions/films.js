import * as types from '../constants/actions';

export const updateSearchResults = (data) => ({
    type: types.updateSearchResults,
    data: JSON.parse(data)
});
