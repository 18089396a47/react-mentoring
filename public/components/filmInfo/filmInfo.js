import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import './filmInforStyles.styl';
import SearchMovieField from '../searchMovieField/searchMovieField';
import InfoPanel from '../infoPanel/infoPanel';
import theMovieDb from 'themoviedb-javascript-library';
import { searchFilmStart } from '../../actions/';
import getYear from '../../helpers/getYear';
import * as commons from '../../constants/commons';

export class FilmInfo extends Component {
    componentWillMount() {
        const path = this.props.history.location.pathname.split('/');

        if (path[1] === 'film') {
            this.props.dispatch(searchFilmStart(path[2], this.props.searchType));
        }
    }

    render() {
        return (
            <div className="film-info">
                <div>
                    <img className="film-info__poster" src={this.props.poster} alt={this.props.show_title} />
                </div>
                <div className="film-info__info-wrapper">
                    <div className="film-info__title-wrapper">
                        <h1 className="film-info__title">{this.props.show_title}</h1>
                        <div className="film-info__rating">{this.props.rating}</div>
                    </div>
                    <div className="film-info__oscar">Oscar-winning Movies</div>
                    <div className="film-info__year-wrapper">
                        <div className="film-info__year">{getYear(this.props.release_year)}</div>
                        <div>{this.props.runtime} min</div>
                    </div>
                    <p className="film-info__summary">{this.props.summary}</p>
                    <div className="film-info__tagline">Tagline: {this.props.tagline}</div>
                    <div className="film-info__companies">Production companies: {this.props.production_companies}</div>
                </div>
            </div>
        );
    }
}

FilmInfo.propTypes = {
    searchType: PropTypes.string.isRequired,
    show_title: PropTypes.string,
    rating: PropTypes.number,
    release_year: PropTypes.string,
    runtime: PropTypes.number,
    summary: PropTypes.string,
    tagline: PropTypes.string,
    production_companies: PropTypes.string,
    poster: PropTypes.string
};

const mapStateToProps = (state) => {
    return {
        searchType: state.search.searchType,
        show_title: state.filmInfo.title || state.filmInfo.name,
        rating: state.filmInfo.vote_average,
        release_year: state.filmInfo.release_date,
        runtime: state.filmInfo.runtime,
        summary: state.filmInfo.overview,
        poster: theMovieDb.common.getImage({
            size: commons.posterSize,
            file: state.filmInfo.poster_path
        }),
        production_companies: state.filmInfo.production_companies
            .map(company => company.name).join(', '),
        tagline: state.filmInfo.tagline
    };
};

export default withRouter(connect(mapStateToProps)(FilmInfo));
