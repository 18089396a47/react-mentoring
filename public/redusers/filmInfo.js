import * as types from '../constants/actions';

const searchReduser = (state = {}, action) => {
    switch (action.type) {
        case types.searchFilmStart:
            return Object.assign({}, state, {
                isFetching: true
            });
        case types.searchFilmEnd:
            return Object.assign({}, state, {
                isFetching: false
            });
        case types.updateFilmInfo:
            return Object.assign({}, action.data);
        default:
            return state;
    }
};

export default searchReduser;
