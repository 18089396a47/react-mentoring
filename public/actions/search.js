import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import { updateSearchResults } from './films';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const changeSearchInput = (inputValue) => ({
    type: types.changeSearchInput,
    inputValue
});

export const searchQueryEnd = () => ({
    type: types.searchQueryEnd
});

export const searchQueryStart = (inputValue, searchType) => ({
    type: types.searchQueryStart,
    inputValue,
    searchType
});

function* fetchQuery({ inputValue, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.search.getMovie : theMovieDb.search.getTv;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            query: inputValue
        }));

        yield put(searchQueryEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchQueryEnd());
    }
};

function* watchFetchQuery() {
    yield takeLatest(types.searchQueryStart, fetchQuery);
}

export const searchSimilarFilmEnd = () => ({
    type: types.searchSimilarFilmEnd
});

export const searchSimilarFilmStart = (id) => ({
    type: types.searchSimilarFilmStart,
    id
});

function* fetchSimilarFilms({ id }) {
    try {
        const response = yield call(() => promiseWrapper(theMovieDb.movies.getSimilarMovies, {
            id
        }));

        yield put(searchSimilarFilmEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchSimilarFilmEnd());
    }
};

function* watchFetchSimilarFilms() {
    yield takeLatest(types.searchSimilarFilmStart, fetchSimilarFilms);
}

export function* searchSaga() {
    yield all([
        watchFetchSimilarFilms(),
        watchFetchQuery()
    ]);
}

export const changeSearchType = (searchType) => ({
    type: types.changeSearchType,
    searchType
});
