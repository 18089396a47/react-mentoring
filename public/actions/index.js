import { all } from 'redux-saga/effects';

import { filmInfoSaga } from './filmInfo';
import { searchSaga } from './search';

export default function* rootSaga() {
    yield all([
        filmInfoSaga(),
        searchSaga()
    ]);
}

export * from './search';
export * from './sortCriteria';
export * from './filmInfo';
export * from './films';
