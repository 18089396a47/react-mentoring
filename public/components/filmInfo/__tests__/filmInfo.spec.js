import React from 'react';
import renderer from 'react-test-renderer';
import { FilmInfo } from '../filmInfo';
import { searchFilmStart } from '../../../actions/';

const mockDigest = jest.fn();

jest.mock('../../searchMovieField/searchMovieField', () => 'SearchMovieField');
jest.mock('../../infoPanel/infoPanel', () => 'InfoPanel');
jest.mock('../../../helpers/getYear', () => jest.fn());
jest.mock('../../../actions/index', () => ({
    searchFilmStart: jest.fn()
}));

let props;

describe('FilmInfo component', () => {
    beforeEach(() => {
        props = {
            dispatch: mockDigest,
            searchType: 'searchType',
            show_title: 'showTitle',
            release_year: 'release_year',
            category: 'category',
            id: 111,
            poster: 'poster',
            rating: 222,
            runtime: 333,
            summary: 'summary',
            tagline: 'tagline',
            production_companies: 'productionCompanies'
        };
    });

    it('FilmInfo should be shown correctly', () => {
        props.history = {
            location: {
                pathname:'/'
            }
        };
        const tree = renderer.create(
            <FilmInfo {...props} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('FilmInfo should call \'searchFilmStart\' redux actions on initializing when specific url', () => {
        props = {
            dispatch: mockDigest,
            searchType: 'searchType',
            history: {
                location: {
                    pathname:'/film/111'
                }
            }
        };
        const tree = shallow(<FilmInfo {...props} />);
        const link = tree.find('.film-item__title-link');

        expect(mockDigest).toHaveBeenCalled();
        expect(searchFilmStart).toHaveBeenCalledWith('111', 'searchType');
    });
});
