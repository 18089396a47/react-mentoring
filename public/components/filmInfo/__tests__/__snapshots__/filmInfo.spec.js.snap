// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmInfo component FilmInfo should be shown correctly 1`] = `
<div
  className="film-info"
>
  <div>
    <img
      alt="showTitle"
      className="film-info__poster"
      src="poster"
    />
  </div>
  <div
    className="film-info__info-wrapper"
  >
    <div
      className="film-info__title-wrapper"
    >
      <h1
        className="film-info__title"
      >
        showTitle
      </h1>
      <div
        className="film-info__rating"
      >
        222
      </div>
    </div>
    <div
      className="film-info__oscar"
    >
      Oscar-winning Movies
    </div>
    <div
      className="film-info__year-wrapper"
    >
      <div
        className="film-info__year"
      />
      <div>
        333
         min
      </div>
    </div>
    <p
      className="film-info__summary"
    >
      summary
    </p>
    <div
      className="film-info__tagline"
    >
      Tagline: 
      tagline
    </div>
    <div
      className="film-info__companies"
    >
      Production companies: 
      productionCompanies
    </div>
  </div>
</div>
`;
